### ISSUE TITLE
update for NumPyro 0.3

### ISSUE THREAD

**Issue by fehiepsi** [RMichae1/PyroStudies#1]
I just jump across your [post](https://towardsdatascience.com/improve-your-model-with-missing-data-imputation-with-numpyro-dcb3c3376eff). It is awesome!!

Just an update: in NumPyro 0.3, the functionality of `numpyro.param` has been removed. The block
```python
if target is not None:
    len_impute = numpyro.param("len_impute", np.zeros(len_is_nan.sum()))
else:
    len_impute = numpyro.sample("len_impute", dist.Normal(len_mu[len_nan_idx],  len_sigma[len_nan_idx]))
```
can be replaced by
```
    len_impute = numpyro.sample("len_impute", dist.Normal(len_mu[len_nan_idx],  len_sigma[len_nan_idx]).mask(False))
```
You might also want to use [numpyro.deterministic("mu", logits or probs)](http://num.pyro.ai/en/stable/primitives.html#deterministic) instead of 
```python
if target is None:
        # prediction case
        probs = expit(logits)
        numpyro.sample("probs", dist.Delta(probs))
```
to simplify the prediction code. :)

**Comment by RMichae1** [comment]
Thanks for reaching out - I'm a big fan of your [NumPyro Statistical Rethinking](https://github.com/fehiepsi/rethinking-numpyro) work!

I'll made changes accordingly, thanks for the heads-up. I have adjusted the post as well.
Your replacement entails, that you index on the samples, so that you have to also adjust the mu and sigma before as jnp arrays of the correct shape:

 ```
len_mu = numpyro.sample("p_length_mu", dist.Normal(jnp.ones(len(p_len))*3, 0.01))
len_sigma = abs(numpyro.sample("p_length_sigma", dist.Normal(jnp.ones(len(p_len)), 0.01)))
```
Is there something like a constraints.positive for numpyro sampling? I have used abs so far, but that does not look ideal

Next time I'll also state what numpyro version I am working with :)

**Comment by fehiepsi** [comment]
Hi @RMichae1, I usually use `LogNormal`, `HalfNormal`, or `Exponential` positive-support priors. I think you can use any of them as a prior for `p_length_sigma`.

> adjust the mu and sigma before as jnp arrays of the correct shape

You are right! What do you think about `numpyro.sample("len_impute", dist.Normal(len_mu,  len_sigma).expand([len(len_is_nan)]).mask(False))`? (`len_mu` and `len_sigma` are hyper-parameters so I guess it is better to let them be scalars)

**Comment by RMichae1** [comment]
Thanks @fehiepsi I will go for the positive priors next time.

Your suggestion is good - though
`len_is_nan` is the boolean array of missing values and is of insufficient size (its the complete array including NaNs), what works for me is:
```
len_impute = numpyro.sample("len_impute", 
                                dist.Normal(len_mu, len_sigma).expand([len(len_nan_idx)]).mask(False))
```
such that I go for the length of the missing indices array.

Its good advice to keep hyper-parameters scalars - I had not considered that.

**Comment by fehiepsi** [comment]
You are right! We need to use the length of missing indices there. :)
